src/backend/cuda/kernel/canny.hpp

```diff
@@ -0,0 +1,372 @@
+            float mag1, mag2;
+
+            if (dx>=0) {
+                if (dy>=0) {
+                    if ((dx-dy) >= 0) {
+                        // East of South East
+                        alpha = dy/dx;
+                        mag1  = (1-alpha) * shrdMem[j][i+1] + alpha * shrdMem[j+1][i+1];
+                        mag2  = (1-alpha) * shrdMem[j][i-1] + alpha * shrdMem[j-1][i-1];
+                    } else {
+                        // South of South East
+                        alpha = dx/dy;
+                        mag1  = (1-alpha) * shrdMem[j+1][i] + alpha * shrdMem[j+1][i+1];
+                        mag2  = (1-alpha) * shrdMem[j-1][i] + alpha * shrdMem[j-1][i-1];
+                    }
+                } else {
+                    if ((dx+dy) >= 0) {
+                        // East of North East
+                        alpha = -dy/dx;
+                        mag1  = (1-alpha) * shrdMem[j][i+1] + alpha * shrdMem[j-1][i+1];
+                        mag2  = (1-alpha) * shrdMem[j][i-1] + alpha * shrdMem[j+1][i-1];
+                    } else {
+                        // North of North East
+                        alpha = dx/-dy;
+                        mag1  = (1-alpha) * shrdMem[j-1][i] + alpha * shrdMem[j-1][i+1];
+                        mag2  = (1-alpha) * shrdMem[j+1][i] + alpha * shrdMem[j+1][i-1];
+                    }
+                }
+            } else {
+                if (dy>=0) {
+                    if ((dx+dy) >= 0) {
+                        // South of South West
+                        alpha = -dx/dy;
+                        mag1  = (1-alpha) * shrdMem[j+1][i] + alpha * shrdMem[j+1][i-1];
+                        mag2  = (1-alpha) * shrdMem[j-1][i] + alpha * shrdMem[j-1][i+1];
+                    } else {
+                        // West of South West
+                        alpha = dy/-dx;
+                        mag1  = (1-alpha) * shrdMem[j][i-1] + alpha * shrdMem[j+1][i-1];
+                        mag2  = (1-alpha) * shrdMem[j][i+1] + alpha * shrdMem[j-1][i+1];
+                    }
+                } else {
+                    if ((-dx+dy) >= 0) {
+                        // West of North West
+                        alpha = -dy/dx;
+                        mag1  = (1-alpha) * shrdMem[j][i-1] + alpha * shrdMem[j-1][i-1];
+                        mag2  = (1-alpha) * shrdMem[j][i+1] + alpha * shrdMem[j+1][i+1];
+                    } else {
+                        // North of North West
+                        alpha = dx/-dy;
+                        mag1  = (1-alpha) * shrdMem[j-1][i] + alpha * shrdMem[j-1][i-1];
+                        mag2  = (1-alpha) * shrdMem[j+1][i] + alpha * shrdMem[j+1][i+1];
+                    }
+                }
+            }
+
+            if (cmag>mag1 && cmag>mag2) {
+                out[idx] = cmag > 255 ? EDGE : (unsigned char)cmag;
+            } else {
+                out[idx] = SUPPRESS;
+            }
+        }
+    }
```

Comment:
Same issues as the CPU implementation. Try to reduce code. Try not to ard code 255.